pkg/controllers/customcluster_upgrade.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright Kurator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"regexp"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	corev1 "k8s.io/api/core/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"kurator.dev/kurator/pkg/apis/infra/v1alpha1"
+)
+
+// reconcileUpgrade is responsible for handling the customCluster reconciliation process of cluster upgrading to targetVersion
+func (r *CustomClusterController) reconcileUpgrade(ctx context.Context, customCluster *v1alpha1.CustomCluster, targetVersion string) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	cmd := generateUpgradeManageCMD(targetVersion)
+	// Checks whether the worker node for upgrading already exists. If it does not exist, then create it.
+	workerPod, err1 := r.ensureWorkerPodCreated(ctx, customCluster, CustomClusterUpgradeAction, cmd, generateClusterHostsName(customCluster), generateClusterConfigName(customCluster))
+	if err1 != nil {
+		conditions.MarkFalse(customCluster, v1alpha1.UpgradeCondition, v1alpha1.UpgradeWorkerCreateFailed,
+			clusterv1.ConditionSeverityWarning, "upgrade worker is failed to create %s/%s.", customCluster.Namespace, customCluster.Name)
+		log.Error(err1, "failed to ensure that upgrade WorkerPod is created", "customCluster", customCluster.Name)
+		return ctrl.Result{}, err1
+	}
+
+	// Check the current customCluster status.
+	if customCluster.Status.Phase != v1alpha1.UpgradingPhase {
+		log.Info("phase changes", "prevPhase", customCluster.Status.Phase, "currentPhase", v1alpha1.UpgradingPhase)
+		customCluster.Status.Phase = v1alpha1.UpgradingPhase
+	}
+
+	// Determine the progress of upgrading based on the status of the workerPod.
+	if workerPod.Status.Phase == corev1.PodSucceeded {
+		// Update the cluster-config to ensure that the current cluster-config (provisioned cluster info) represents the cluster after upgrading.
+		if err := r.updateKubeVersion(ctx, customCluster, targetVersion); err != nil {
```

Comment:
why update after succeeded, how does it work without updating before run worker?

Reply:
In kubespray, we can just tell it the desired version and upgrade will process. when the upgrade is successd, we update the recode of provisioned info. In this case, if upgrade is err, we can delete the err one and auto restart the upgrade pod.